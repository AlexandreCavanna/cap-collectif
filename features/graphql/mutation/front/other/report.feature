@comments @other
Feature: Comments

@database
Scenario: Admin GraphQL client wants to report a idea
  Given I am logged in to graphql as admin
  And I send a GraphQL POST request:
  """
  {
    "query": "mutation ToggleFeatureMutation($input: ReportInput!) {
      report(input: $input) {
        report {
          body
          type
        }
      }
    }",
    "variables": {
      "input": {
        "reportableId": "Q29tbWVudDpwcm9wb3NhbENvbW1lbnQx",
        "body": "je suis un spammeur",
        "type": "SPAM"
      }
    }
  }
  """
  Then the JSON response should match:
  """
  {
    "data": {
      "report": {
        "report": {
          "body": "je suis un spammeur",
          "type": "SPAM"
        }
      }
    }
  }
  """

@security
Scenario: Anonymous GraphQL client wants to report a idea
  Given I am logged out
  And I send a GraphQL POST request:
  """
  {
    "query": "mutation ToggleFeatureMutation($input: ReportInput!) {
      report(input: $input) {
        report {
          body
          type
        }
      }
    }",
    "variables": {
      "input": {
        "reportableId": "Q29tbWVudDpwcm9wb3NhbENvbW1lbnQx",
        "body": "je suis un spammeur",
        "type": "SPAM"
      }
    }
  }
  """
  Then the JSON response should match:
  """
  {
    "errors": [
      {"message":"Access denied to this field.","@*@": "@*@"}
    ],
    "data": {
      "report": null
    }
  }
  """
