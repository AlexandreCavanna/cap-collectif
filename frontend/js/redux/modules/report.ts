import type { Exact, Action, Dispatch } from '~/types'
import { UPDATE_ALERT } from '~/constants/AlertConstants'
import FluxDispatcher from '../../dispatchers/AppDispatcher'
import ReportMutation from '~/mutations/ReportMutation'
import { getType } from '~/components/Report/ReportForm'
export type State = {
  readonly currentReportingModal: (number | null | undefined) | (string | null | undefined)
  readonly isLoading: boolean
  readonly elements: Array<(number | null | undefined) | (string | null | undefined)>
}
export type ReportData = {
  status: number
  body: string
}
const initialState: State = {
  currentReportingModal: null,
  isLoading: false,
  elements: [],
}
type OpenModalAction = {
  type: 'report/OPEN_MODAL'
  id: number
}
type CloseModalAction = {
  type: 'report/CLOSE_MODAL'
}
type StartReportingAction = {
  type: 'report/START_LOADING'
}
type StopReportingAction = {
  type: 'report/STOP_LOADING'
}
type AddReportedAction = {
  type: 'report/ADD_REPORTED'
}
export type ReportAction =
  | OpenModalAction
  | CloseModalAction
  | StartReportingAction
  | StopReportingAction
  | AddReportedAction
export const openModal = (id: number): OpenModalAction => ({
  type: 'report/OPEN_MODAL',
  id,
})
export const closeModal = (): CloseModalAction => ({
  type: 'report/CLOSE_MODAL',
})

const startLoading = (): StartReportingAction => ({
  type: 'report/START_LOADING',
})

const stopLoading = (): StopReportingAction => ({
  type: 'report/STOP_LOADING',
})

const addReported = (): AddReportedAction => ({
  type: 'report/ADD_REPORTED',
})

export const submitReport = async (
  reportableId: string,
  report: ReportData,
  dispatch: Dispatch,
  successMessage: string,
) => {
  dispatch(startLoading())
  await ReportMutation.commit({
    input: {
      reportableId,
      type: getType(report.status.toString()),
      body: report.body,
    },
  })
  dispatch(addReported())
  dispatch(stopLoading())
  dispatch(closeModal())
  FluxDispatcher.dispatch({
    actionType: UPDATE_ALERT,
    alert: {
      bsStyle: 'success',
      content: successMessage,
    },
  })
}
export const reducer = (state: State = initialState, action: Action): Exact<State> => {
  switch (action.type) {
    case 'report/START_LOADING':
      return { ...state, isLoading: true }

    case 'report/STOP_LOADING':
      return { ...state, isLoading: false }

    case 'report/OPEN_MODAL':
      return { ...state, currentReportingModal: action.id }

    case 'report/CLOSE_MODAL':
      return { ...state, currentReportingModal: null }

    case 'report/ADD_REPORTED': {
      return { ...state, elements: [...state.elements, state.currentReportingModal] }
    }

    default:
      return state
  }
}
