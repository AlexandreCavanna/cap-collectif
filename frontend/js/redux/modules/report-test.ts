/* eslint-env jest */
import { reducer } from './report'

const defaultInitialState = {
  currentReportingModal: null,
  isLoading: false,
  elements: [],
}
describe('Report Reducer', () => {
  it('Should handle OPEN_MODAL', () => {
    const newState = reducer(defaultInitialState, {
      type: 'report/OPEN_MODAL',
      id: 1,
    })
    expect(newState).toMatchSnapshot()
  })
  it('Should handle CLOSE_MODAL', () => {
    const initialState = { ...defaultInitialState, currentReportingModal: 'lalaa' }
    const newState = reducer(initialState, {
      type: 'report/CLOSE_MODAL',
    })
    expect(newState).toMatchSnapshot()
  })
  it('Should handle START_LOADING', () => {
    const newState = reducer(defaultInitialState, {
      type: 'report/START_LOADING',
    })
    expect(newState).toMatchSnapshot()
  })
  it('Should handle STOP_LOADING', () => {
    const initialState = { ...defaultInitialState, isLoading: true }
    const newState = reducer(initialState, {
      type: 'report/STOP_LOADING',
    })
    expect(newState).toMatchSnapshot()
  })
  it('Should handle ADD_REPORTED', () => {
    const initialState = { ...defaultInitialState, currentReportingModal: 'foo', elements: ['bar'] }
    const newState = reducer(initialState, {
      type: 'report/ADD_REPORTED',
    })
    expect(newState).toMatchSnapshot()
  })
})
