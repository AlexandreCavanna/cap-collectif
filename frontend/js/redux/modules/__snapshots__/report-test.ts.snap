// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Report Reducer Should handle ADD_REPORTED 1`] = `
Object {
  "currentReportingModal": "foo",
  "elements": Array [
    "bar",
    "foo",
  ],
  "isLoading": false,
}
`;

exports[`Report Reducer Should handle CLOSE_MODAL 1`] = `
Object {
  "currentReportingModal": null,
  "elements": Array [],
  "isLoading": false,
}
`;

exports[`Report Reducer Should handle OPEN_MODAL 1`] = `
Object {
  "currentReportingModal": 1,
  "elements": Array [],
  "isLoading": false,
}
`;

exports[`Report Reducer Should handle START_LOADING 1`] = `
Object {
  "currentReportingModal": null,
  "elements": Array [],
  "isLoading": true,
}
`;

exports[`Report Reducer Should handle STOP_LOADING 1`] = `
Object {
  "currentReportingModal": null,
  "elements": Array [],
  "isLoading": false,
}
`;
